gnmi/gnmi.go

```diff
@@ -0,0 +1,152 @@
+package gnmi
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	gpb "github.com/openconfig/gnmi/proto/gnmi"
+	"github.com/openconfig/magna/lwotgtelem"
+	"github.com/openconfig/magna/lwotgtelem/gnmit"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/reflection"
+	"k8s.io/klog/v2"
+)
+
+type GNMI struct {
+	cfg *Config
+
+	lis      net.Listener
+	srv      *grpc.Server
+	metricCh chan *pmetric.Metrics
+	telemSrv *lwotgtelem.Server
+}
+
+func NewGNMIExporter(cfg *Config) (*GNMI, error) {
+	telemSrv, err := lwotgtelem.New(context.Background(), cfg.TargetName)
+	if err != nil {
+		return nil, err
+	}
+
+	gnmiLis, err := net.Listen("tcp", fmt.Sprintf("%s", cfg.Addr))
+	if err != nil {
+		klog.Exitf("cannot listen on %s, err: %v", cfg.Addr, err)
+	}
+
+	var opts []grpc.ServerOption
+
+	if cfg.CertFile != "" {
+		creds, err := credentials.NewServerTLSFromFile(cfg.CertFile, cfg.KeyFile)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create gNMI credentials, %v", err)
+		}
+		opts = append(opts, grpc.Creds(creds))
+	}
+
+	return &GNMI{
+		cfg:      cfg,
+		metricCh: make(chan *pmetric.Metrics, cfg.BufferSize),
+		lis:      gnmiLis,
+		srv:      grpc.NewServer(opts...),
+		telemSrv: telemSrv,
+	}, nil
+}
+
+func (g *GNMI) Start(_ context.Context, _ component.Host) error {
+	fmt.Println("starting")
+	reflection.Register(g.srv)
+	gpb.RegisterGNMIServer(g.srv, g.telemSrv.GNMIServer)
+	go g.srv.Serve(g.lis)
+	g.telemSrv.AddTask(gnmit.Task{
+		Run: g.handleMetrics,
+	})
+	fmt.Println("started")
+	return nil
+}
+
+func (g *GNMI) Stop(_ context.Context) error {
+	close(g.metricCh)
+	g.srv.GracefulStop()
+	return nil
+}
+
+func (g *GNMI) storeMetric(ctx context.Context, md pmetric.Metrics) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		g.metricCh <- &md
+	}
+	return nil
+}
+
+func (g *GNMI) handleMetrics(_ gnmit.Queue, updateFn gnmit.UpdateFn, target string, cleanup func()) error {
```

Comment:
Can we add a comment for what this function handles -- it'd be good to understand what the looping is trying to achieve before reading through.